packages/webdriverio/src/utils/interception/devtools.js

```diff
@@ -0,0 +1,182 @@
+import fs from 'fs'
+import path from 'path'
+import atob from 'atob'
+import minimatch from 'minimatch'
+import { canAccess } from '@wdio/utils'
+
+import { containsObject } from '..'
+import { ERROR_REASON } from '../../constants'
+
+export default class DevtoolsInterception {
+    constructor (url, filterOptions = {}) {
+        this.url = url
+        this.filterOptions = filterOptions
+        this.respondOverwrites = []
+        this.matches = []
+    }
+
+    static handleRequestInterception (client, mocks) {
+        return async (event) => {
+            const responseHeaders = event.responseHeaders.reduce((headers, { name, value }) => {
+                headers[name] = value
+                return headers
+            }, {})
+            const { requestId, request } = event
+
+            for (const mock of mocks) {
+                /**
+                 * match mock url
+                 */
+                if (!minimatch(request.url, mock.url)) {
+                    continue
+                }
+
+                /**
+                 * match filter options
+                 */
+                if (
+                    /**
+                     * HTTP method
+                     */
+                    (mock.filterOptions.method && mock.filterOptions.method.toLowerCase() !== request.method.toLowerCase()) ||
+                    (mock.filterOptions.headers && containsObject(responseHeaders, mock.filterOptions.headers))
+                ) {
+                    continue
+                }
+
+                const { body, base64Encoded } = await client.send(
+                    'Fetch.getResponseBody',
+                    { requestId }
+                )
+
+                request.body = base64Encoded ? atob(body) : body
+                request.body = responseHeaders['Content-Type'] && responseHeaders['Content-Type'].includes('application/json')
+                    ? JSON.parse(request.body)
+                    : request.body
+                mock.matches.push(request)
+
+                /**
+                 * no stubbing if no overwrites were defined
+                 */
+                if (mock.respondOverwrites.length === 0) {
+                    continue
+                }
+
+                const { errorReason, overwrite, params } = mock.respondOverwrites[0].sticky
+                    ? mock.respondOverwrites[0]
+                    : mock.respondOverwrites.shift()
+
+                /**
+                 * when response is modified
+                 */
+                if (overwrite) {
+                    let body = overwrite
```

Comment:
there is also `body` variable declared at line 47. I know they are not conflicting but might be confusing.